### DISCUSSION TITLE
General questions: outputs & SVM case

### DISCUSSION THREAD

**Discussion by seralouk** [OpenNFT/OpenNFT#114] [Q&A]
Hi there,

I am posting here some questions. I have successfully installed the OpenNFT software and I have run the tutorials (PSC cont, SVM). I have also read your NeuroImage manuscript.

**Questions:**

My goal: I plan to use the software for a new rt-NF-fMRI experiment. The main idea is to train an SVM on localizer fMRI runs and then use rt-neurofeedback using OpenNFT. During the neurofeedback runs, 2 conditions will be used (baseline, task). The feedback from the SVM mask should be used to modify the difficulty of the task condition.

1. How can I make sense of all the saved outputs? Is there a manual? Where should someone who would like to use the software but has little experimental fMRI setup experience start?
<img width="408" alt="image" src="https://user-images.githubusercontent.com/25902670/124477263-1e708e00-dda4-11eb-8ae6-bc89cd9c350b.png">

2. In the SVM toy example, a) why is not the SVM feedback value printed in my console/terminal as in the tutorial (I also tried to manually add `USE_UDP_FEEDBACK = True` in `config.py`, but no luck)? b) Is this feedback the dot product of the SVM weights (W1 folder, `weights_parietal.nii`) with the activity estimated from the same SVM binary mask (ROI folder, `mask_parietal.nii`) ? So, the binary mask is used to just extract the voxel-wise BOLD signal and then the dot is estimated using the weights?

3. The norm ROI signal is the output of the sigmoid transformation applied on the dot product of the SVM weights & brain activity from the mask (`prediction_of_scan_1 = sigmoid(w.T*signals_scan_1)`)? What is the light blue line in this case? I would only expect to have the dark blue.

![image](https://user-images.githubusercontent.com/25902670/124478298-575d3280-dda5-11eb-83a9-d9e66f9ff0ec.png)

Thank you in advance for your help & the amazing software.

**Comment by tiborauer** [comment]
Ad 2.: As @ykoush suggested you should use our [mock UDP receiver](https://github.com/OpenNFT/OpenNFT/blob/master/tests/testUDPReceiver.py).

Ad 3.: The light blue is the feedback value as sent. It is zero during the control periods.

**Reply by seralouk** [reply]
Thank you for your answer. For Ad 3., during the control periods, what does the dark blue line (Norm ROI) represent?

**Reply by tiborauer** [reply]
The dark blue is the calculated "activation" value, which is transformed into the feedback value(light blue). In the case of a single-ROI experiment with ROI operation "mean(norm_percValues)" (seen above the MC plot), they are the same (i.e. mean of a single value).

**Reply by seralouk** [reply]
Got it! And the dark blue is the transformed values after applying the sigmoid on the dot product of the SVM weights with the brain activity?

**Reply by tiborauer** [reply]
Correct.

**Reply by seralouk** [reply]
Thank you. May I ask how exactly I could use your mock UDP receiver in order to get the SVM feedback printed on the terminal or extracted/saved in an external file? I have tried to add the following lines in my `config.py`:
```
USE_UDP_FEEDBACK = True
UDP_FEEDBACK_IP = "127.0.0.1"
UDP_FEEDBACK_PORT = 1234
UDP_FEEDBACK_CONTROLCHAR = '#'
```
after seeing 
```
self.udpSender = Udp(
  IP=config.UDP_FEEDBACK_IP,
  port=config.UDP_FEEDBACK_PORT,
  control_signal=config.UDP_FEEDBACK_CONTROLCHAR,
  encoding='UTF-8'
```

in `opennft.py`. Sorry for this basic question but I am not familiar (yet) with the UDP.

**Comment by tiborauer** [comment]
First, you do not need to add anything to `config.py` but to the `.ini` file or configure the setup via the GUI. The default settings correspond to the mock UDP setup, so you only need to enable UPD (and "Send condition").

Second, you need to run the `testUDPReceiver.py` in a separate Anaconda terminal (as it is a separate script emulating a feedback visualizer).
1. Open a new Anaconda Terminal
2. Activate the opennft virtual environment
3. Navigate to the opennft/test folder (in the terminal)
4. `python testUDPReceiver.py`

(5. Your computer may ask you to allow the communication via the firewall.)

**Reply by seralouk** [reply]
Thank you a lot for the detailed answer. I have made the necessary changes (enabled UPD and "Send condition") but I get a connection error even when all firewalls are off on my MacBook. 

Any idea how to overcome this?

```
=============
Matlab time stamp t6!
=============

2021-07-09 10:09:07 | INFO      | opennft:callMainLoopIteration - Sending by UDP - instrValue =
2021-07-09 10:09:07 | INFO      | opennft:callMainLoopIteration - Call iteration for file "001_000007_000006.dcm"
TIMING: 5 iter - PREPROC MC: 3.478519e-01 s - SMOOTH: 1.109055e-02 s - IGLM: 2.424289e-02 s
2021-07-09 10:09:07 | INFO      | opennft:callMainLoopIteration -   preprocess fMRI Volume: 0.40 s
2021-07-09 10:09:07 | INFO      | opennft:callMainLoopIteration -   preprocess signal: 6.09 ms
2021-07-09 10:09:07 | INFO      | opennft:callMainLoopIteration - Sending by UDP - dispValue = 0.0
Traceback (most recent call last):
  File "/Users/loukas/Desktop/opennft-venv/lib/python3.7/site-packages/opennft/opennft.py", line 912, in callMainLoopIteration
    self.udpSender.send_data(self.displayData['dispValue'])
  File "/Users/loukas/Desktop/opennft-venv/lib/python3.7/site-packages/pyniexp/connection.py", line 150, in send_data
    else: self._socket.send(self._formats[type(d).__name__]['encode'](d))
ConnectionRefusedError: [Errno 61] Connection refused
```

**Reply by tiborauer** [reply]
Was there any output or error message on the Anadonca Terminal running the testUDPReceiver?

**Reply by seralouk** [reply]
> Was there any output or error message on the Anadonca Terminal running the testUDPReceiver?

No output at all.  After opening a new terminal and activating the virtual environment, `python /OpenNFT/tests/testUDPReceiver.py` does not return any output or error. To summarize my steps in case I do something wrong: After activating the virtual environment I launch `opennft`. I setup the SVM ini and enable the UPD and "send condition" > setup > start.  The output of this terminal is like I posted above, and at the same time, in another terminal with the environment activated, I use `python OpenNFT/tests/testUDPReceiver.py` but it does not return anything. After inspecting the code, `receiver.connect_for_receiving()` line gets stuck...

**Reply by tiborauer** [reply]
OK. I should have been more clear. It is crucial that you start testUDPReceiver before pressing "Setup", when the UDP connection is established. You should see corresponding messages/outputs on both ends.

**Reply by seralouk** [reply]
> OK. I should have been more clear. It is crucial that you start testUDPReceiver before pressing "Setup", when the UDP connection is established. You should see corresponding messages/outputs on both ends.

Got it! thanks!
